### PR TITLE
Add bin and lib directories

### LINKED ISSUE #1753: Create asdf plugin
Create a plugin similar to https://github.com/software-mansion/asdf-scarb to support installation through the version manager

I created a repo for it here: https://github.com/dojoengine/asdf-dojo

### PR DESCRIPTION
**Pull request by Jonatan-Chaverri**
Closes https://github.com/dojoengine/dojo/issues/1753

### CODE REVIEW

**Comment by Jonatan-Chaverri** [comment]
Creating initial commit. Some few questions I have:

1. Since I'm using dojoup to install dojo, how does this tool handles the versioning part? How can a user remove a version and install a different one? Can I have multiple versions on my system? How to switch between them?
2. I noticed that I can specify any version I want to this tool and it will still say that the version was installed successfully. Attaching screenshot
Screenshot 2024-04-14 at 3 41 50 PM

**Comment by kariy** [comment]
> 1. Since I'm using dojoup to install dojo, how does this tool handles the versioning part? How can a user remove a version and install a different one? Can I have multiple versions on my system? How to switch between them?

the `dojoup` script doesn't provide feature for managing different installation versions. if you have already done an installation, subsequent installations using `dojoup` (no matter the install method) would overwrite the previous ones. so, installing using `dojoup` will always result in a single version. even if you somehow have different versions installed, `dojoup` doesn't provide a way to switch between them.

> 2. I noticed that I can specify any version I want to this tool and it will still say that the version was installed successfully. Attaching screenshot

good point! `dojoup` still doesn't handle invalid version properly right now. ideally, it should give a proper error message and abort the installation process immediately. cc @glihm @tarrencev 

hope that answers your questions!

**Comment by glihm** [comment]
@kariy mentioned a good point and I wasn't sure about the `asdf` way to interact with scripts. Reading a bit the documentation, it seems you'll be better at doing specified script for each action required to be done.

Also @Jonatan-Chaverri you can take [scarb](https://github.com/software-mansion/asdf-scarb) as an example.

> good point! dojoup still doesn't handle invalid version properly right now. ideally, it should give a proper error message and abort the installation process immediately. cc @glihm @tarrencev

Good catch yeah, this should be corrected. Will open an issue to ensure this is tracked.
(https://github.com/dojoengine/dojo/issues/1831)

**Comment by Jonatan-Chaverri** [comment]
@glihm it seems to be working now (the installing and list versions scripts). However, I get an error when trying to use katana
```bash
$katana --version
No version is set for command katana
Consider adding one of the following versions in your config file at
dojo 0.5.1
dojo 0.6.0
```
Do you know how to fix this or what might be happening? This start to occur after installing multiple versions of dojo.

File: bin/list-all [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    echo "Error: curl is not installed." >&2
+    exit 1
+fi
+
+if ! command -v git &>/dev/null; then
+    echo "Error: git is not installed." >&2
+    exit 1
+fi
+
+all_versions=$(git ls-remote --tags --refs "$GH_REPO" |
```
Comment (glihm) [inline bin/list-all:18]: Just seen that scarb also used `git`. Perhaps using GitHub API it should be only relying on `curl`? But as a contributor mentioned, the GitHub API is paginated. So a bit of logic implied. This one is not paginated @Jonatan-Chaverri?

**Review by glihm — COMMENTED** [review]
What's the actual path you have @Jonatan-Chaverri when doing the install? Is a path different for each version? Or every bin is sent to the the same path?

File: bin/list-all (hunk 1 above)
Comment (Jonatan-Chaverri) [reply bin/list-all:18]: No, git command will return all the tags with no pagination.

**Comment by Jonatan-Chaverri** [comment]
@glihm @kariy This PR is ready for review now. I tested it and it is working fine. Issue was that I needed to choose a version with `asdf global dojo <version>`. 
Here is a screenshot of it working
Screenshot 2024-04-22 at 8 34 53 PM

**Comment by Jonatan-Chaverri** [comment]
Let me know if I should configure some github workflows also. Not sure if hardware is available or if my user can do that.

**Review by kariy — COMMENTED** [review]
nice work! but i just realized that they have a [template](https://github.com/asdf-vm/asdf-plugin-template/tree/main) for asdf-plugins. have you seen it?. seems like a no brainer to just use that one.

**Comment by glihm** [comment]
> nice work! but i just realized that they have a [template](https://github.com/asdf-vm/asdf-plugin-template/tree/main) for asdf-plugins. have you seen it?. seems like a no brainer to just use that one.

Yep that's a good point.

>  Let me know if I should configure some github workflows also. Not sure if hardware is available or if my user can do that.

In the template it seems they also already give a good starting point, you can definitely add a `.github` with the template files + adapt them to our plugin. Doing so, we will ensure that the plugin is functional on the CI and push confidently new updates. 👍

**Comment by Jonatan-Chaverri** [comment]
I haven't see that before... thank you @kariy. I will close this and create a new PR with that template.

**Pull request closed without merging**